cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,483 @@
+	return f.addDropletsFunc(ctx, fID)
+}
+
+// RemoveDroplets from a Firewall.
+func (f *fakeFirewallService) RemoveDroplets(ctx context.Context, fID string, dropletIDs ...int) (*godo.Response, error) {
+	return f.removeDropletsFunc(ctx, fID, dropletIDs...)
+}
+
+// AddTags to a Firewall.
+func (f *fakeFirewallService) AddTags(ctx context.Context, fID string, tags ...string) (*godo.Response, error) {
+	return f.addTagsFunc(ctx, fID, tags...)
+}
+
+// RemoveTags from a Firewall.
+func (f *fakeFirewallService) RemoveTags(ctx context.Context, fID string, tags ...string) (*godo.Response, error) {
+	return f.removeTagsFunc(ctx, fID, tags...)
+}
+
+// AddRules to a Firewall.
+func (f *fakeFirewallService) AddRules(ctx context.Context, fID string, rr *godo.FirewallRulesRequest) (*godo.Response, error) {
+	return f.addRulesFunc(ctx, fID, rr)
+}
+
+// RemoveRules from a Firewall.
+func (f *fakeFirewallService) RemoveRules(ctx context.Context, fID string, rr *godo.FirewallRulesRequest) (*godo.Response, error) {
+	return f.removeRulesFunc(ctx, fID, rr)
+}
+
+func newFakeFirewall(workerFirewallName string, inboundRule godo.InboundRule) *godo.Firewall {
+	return &godo.Firewall{
+		ID:           "123",
+		Name:         workerFirewallName,
+		InboundRules: []godo.InboundRule{inboundRule},
+	}
+}
+
+func newFakeFirewallManagerOp(client *godo.Client, cache firewallCache) *firewallManagerOp {
+	return &firewallManagerOp{
+		client:             client,
+		fwCache:            cache,
+		workerFirewallName: fakeWorkerFWName,
+		workerFirewallTags: fakeWorkerFWTags,
+	}
+}
+
+func newFakeFirewallCache(inboundRule godo.InboundRule) firewallCache {
+	return firewallCache{
+		mu: new(sync.RWMutex),
+		firewall: &godo.Firewall{
+			ID:           "123",
+			Name:         fakeWorkerFWName,
+			InboundRules: []godo.InboundRule{inboundRule},
+		},
+	}
+}
+
+func newFakeFirewallCacheEmpty() firewallCache {
+	return firewallCache{
+		mu: new(sync.RWMutex),
+	}
+}
+
+func newFakeGodoClient(fakeFirewall *fakeFirewallService) *godo.Client {
+	return &godo.Client{
+		Firewalls: fakeFirewall,
+	}
+}
+
+func TestFirewallController_Get(t *testing.T) {
+	testcases := []struct {
+		name                         string
+		fwCache                      firewallCache
+		expectedGodoFirewallGetResp  func(context.Context, string) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallListResp func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error)
+		expectedError                error
+		expectedFirewall             *godo.Firewall
+	}{
+		{
+			name:    "return error when error on GET firewall by ID",
+			fwCache: newFakeFirewallCache(fakeInboundRule),
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errFailedToRetrieveFirewallByID
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
```

Comment:
The list method is not supposed to be called, is it?

If it is not, then my suggestion would be to _not_ define `expectedGodoFirewallListResp` so that it stays `nil` and makes the test fail hard if any unexpected stub methods were invoked.

In that case, could you please double-check all other test cases as well? We should aim for only stubbing out methods we expect to call.

Reply:
This should be the only one. Just double checked. `return nil, nil, nil` makes this pointless. Nice catch! 👍